teku/src/main/java/tech/pegasys/teku/cli/util/YamlConfigFileParamsProvider.java

```diff
@@ -26,41 +26,83 @@
 import java.util.Collection;
 import java.util.Collections;
 import java.util.HashSet;
+import java.util.List;
 import java.util.Map;
+import java.util.Map.Entry;
 import java.util.Optional;
 import java.util.Set;
 import java.util.TreeSet;
 import java.util.stream.Collectors;
 import picocli.CommandLine;
-import picocli.CommandLine.IDefaultValueProvider;
-import picocli.CommandLine.Model.ArgSpec;
 import picocli.CommandLine.Model.CommandSpec;
 import picocli.CommandLine.Model.OptionSpec;
 import picocli.CommandLine.ParameterException;
 
-public class YamlConfigFileDefaultProvider implements IDefaultValueProvider {
-
+public class YamlConfigFileParamsProvider implements AdditionalParamsProvider {
   private final CommandLine commandLine;
   private final File configFile;
   // this will be initialized on fist call of defaultValue by PicoCLI parseArgs
   private Map<String, Object> result;
 
-  public YamlConfigFileDefaultProvider(final CommandLine commandLine, final File configFile) {
+  public YamlConfigFileParamsProvider(final CommandLine commandLine, final File configFile) {
     this.commandLine = commandLine;
     this.configFile = configFile;
   }
 
   @Override
-  public String defaultValue(final ArgSpec argSpec) {
+  public Map<String, String> getAdditionalParams(final List<OptionSpec> potentialParams) {
     if (result == null) {
       result = loadConfigurationFromFile();
       checkConfigurationValidity(result == null || result.isEmpty());
       checkUnknownOptions(result);
     }
 
-    // only options can be used in config because a name is needed for the key
-    // so we skip default for positional params
-    return argSpec.isOption() ? getConfigurationValue(((OptionSpec) argSpec)) : null;
+    return result.entrySet().stream()
+        .map(yamlEntry -> Map.entry(translateKey(yamlEntry.getKey()), yamlEntry.getValue()))
+        .map(yamlEntry -> mapParam(potentialParams, yamlEntry))
+        .filter(Optional::isPresent)
+        .map(Optional::get)
+        .collect(Collectors.toMap(Entry::getKey, Entry::getValue));
+  }
+
+  private String translateKey(final String yamlKey) {
+    return "--" + yamlKey;
```

Comment:
I don't think this correctly handles options that use a single `-` which we have a few of, mostly in subcommands.

Probably also applies to the env var handler too though apparently we never handled it properly there...